DEVX-2067: First draft of validation library for App Builder templates

AmyJZhao opened the pull request:
## Description

https://jira.corp.adobe.com/browse/DEVX-2067
https://wiki.corp.adobe.com/display/dxdevx/Validation+checks

NPM CLI tool that runs checks on metadata of NPM packages containing App Builder templates. 

```bash
tv run-checks <link to NPM package>
tv run-checks https://www.npmjs.com/package/@adobe/generator-app-asset-compute
```

## Motivation and Context

Meant to be integrated in template submission workflow in GitHub Actions for https://github.com/adobe/aio-template-submission

## How Has This Been Tested?

Clone the repository locally and `cd` into folder.

```bash
npm init
npm test
```

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

lenaorobei reviewed (commented):
Let's add ESLint and some unit tests to cover this functionality. Please add Adobe copyright as well.

It makes sense to sync with @VladimirZaets and see what standard he's using for the API code.

lenaorobei commented on checks/nameContainsKeyword.js:
```diff
@@ -0,0 +1,26 @@
+async function nameContainsKeyword(metadataJSON) {
```
The idea behind the "keywords" check was to verify that the "keywords" property contains "aio-app-builder-template". As far as I'm concerned, we agreed not to have any restrictions about the template name.

See https://docs.npmjs.com/cli/v8/configuring-npm/package-json#keywords

AmyJZhao replied:
Hi Olena, I have fixed this function.

AmyJZhao commented:
> Let's add ESLint and some unit tests to cover this functionality. Please add Adobe copyright as well.
> 
> It makes sense to sync with @VladimirZaets and see what standard he's using for the API code.

Hi Olena, I've added ESLint and unit tests.

> Let's add ESLint and some unit tests to cover this functionality. Please add Adobe copyright as well.
> 
> It makes sense to sync with @VladimirZaets and see what standard he's using for the API code.

Hi @lenaorobei , I have added ESLint, unit tests, and Adobe copyright headers.

lenaorobei commented on package.json:
```diff
@@ -0,0 +1,40 @@
+{
+  "name": "aio-lib-template-validation",
```
Looks like "author" is missing.

lenaorobei commented on src/checks/containsKeyword.js:
```diff
@@ -0,0 +1,42 @@
+
+module.exports = {
+    method: containsKeyword,
+    description: 'NPM package should have keyword aio-app-builder-template',
```
```suggestion
    description: 'NPM package should have keyword "aio-app-builder-template"',
```

AmyJZhao replied:
Done @lenaorobei [b8baff3](https://github.com/adobe/aio-lib-template-validation/pull/1/commits/b8baff3e831abda3467d5a1c1a156a55d0e5e013)

roribio commented on src/checks/checkNodeVersion.js:
```diff
@@ -0,0 +1,39 @@
+*/
+
+async function checkNodeVersion(metadataJSON) {
+    if ('_nodeVersion' in metadataJSON) {
```
If the `metadataJSON` object is `null`, I believe this `if` will throw an error. We should catch it or throw it. Please check if there are any other instances of this in the other functions.

AmyJZhao replied:
Done. If function detects metadataJSON is null, it returns a 'fail' status

AmyJZhao commented on package.json:
```diff
@@ -0,0 +1,40 @@
+    "type": "git",
+    "url": "git+https://github.com/adobe/aio-lib-template-validation.git"
+  },
+  "author": "Amy Zhao <azhao@adobe.com>",
```
Done @lenaorobei [b8baff3](https://github.com/adobe/aio-lib-template-validation/pull/1/commits/b8baff3e831abda3467d5a1c1a156a55d0e5e013)

roribio commented on src/checks/checkNodeVersion.js:
```diff
@@ -0,0 +1,39 @@
+async function checkNodeVersion(metadataJSON) {
+    if ('_nodeVersion' in metadataJSON) {
+        var nodeVersion = metadataJSON._nodeVersion
+        const regex14 = new RegExp('^14(.[0-9]d*)*')
+        const regex16 = new RegExp('^16(.[0-9]d*)*')
```
This is fine for now, but what happens when the supported versions change? We'll surely need to actively maintain these lines.

AmyJZhao replied:
@roribio I updated checkNodeVersion to have a const array of supported versions, and the function will loop through the array and run a regex on each.

roribio commented on tests/checkNodeVersion.test.js:
```diff
@@ -0,0 +1,73 @@
+const { describe, it, expect } = require('@jest/globals')
+const checkNodeVersion = require('../src/checks/checkNodeVersion')
+
+describe(`checks node version ^14 || ^16`, function () {
+    it('returns status: pass', async () => {
```
I would add an additional test condition to all tests where the JSON object is `null`.

AmyJZhao replied:
Done.

roribio reviewed (commented):
Good work, Amy 👍 I leave my comments for your consideration.

roribio reviewed (approved):
Looks good 👍 Good work, Amy 💯 

I'm approving to not hold this back longer, but in the future consider encapsulating this logic https://github.com/adobe/aio-lib-template-validation/pull/1/files#diff-4e4b6209cc46eca39bbe3f2e391cb3f3324b21e308234f5ce8ced92cf2863fa0R18-R23 in it's own function to keep code DRY.

VladimirZaets commented on src/checks/checkNodeVersion.js:
```diff
@@ -0,0 +1,51 @@
+    if (!metadataJSON) {
+        return {
+            message: 'Error: metadataJSON is null',
+            status: 'fail',
```
Status "fail" isn't the best option in this case. In general "fail" isn't a good word, it doesn't follow any practice and can be applied only in UI but not programmatically. In this case, I propose to use boolean value and instead of status property, use "error".  
So, the code should look like: 
```
return {
    message: 'Error: metadataJSON is null',
    error: true
}
```
A best practice if you needed to introduce some special words, like 'fail', 'pass' etc is to introduce constants that will be sharable between modules, and in most cases is good to represent them as enum type. This will give the possibility to not depend exactly on the string, we will depend on the entity.  It should be something like:
```
module.exports = {
    FAIL: 0,
    PASS: 1,
    SOME_OTHER_STATUS: 2
    ...
}
```

but again, we need to try to avoid it were it possible. If we have just 2 statuses the better choice use the boolean value.


For errors, you can use the new Error('Message').toString() construction if you want have error as a string. For simplifying debugging you also could add stack trace by new Error('Message').stack call. 

if (result.status === 'pass') {
                    return true
                }

AmyJZhao replied:
I'll think about this as a possible area to refactor. I get your point about the statuses, and I'll probably change them into boolean/enum at some point. However, on the subject of errors, I don't think checks failing should throw an error for the entire test suite, which is why I lumped bad inputs and missing keys into the 'fail' status alongside the check condition.

VladimirZaets commented on src/commands/check.js:
```diff
@@ -0,0 +1,35 @@
+governing permissions and limitations under the License.
+*/
+
+var exec = require('sync-exec')
```
using "var" is an outdated approach, please use let or const

AmyJZhao replied:
Done. replaced var with let

VladimirZaets commented on src/commands/check.js:
```diff
@@ -0,0 +1,35 @@
+    var templateMetadata = await getTemplateMetadata(packageName)
+    var templateJSON = JSON.parse(templateMetadata)
+    // run all checks on metadata JSON object
+    await runChecks(templateJSON)
```
Is it any reason to wait for execution of a function that doesn't return anything?

AmyJZhao replied:
I think runChecks has to be an async function because I have the checks run asynchronously within runChecks. Since runChecks has to be an asynchronous function, same goes for check() 
When I turn check() and runChecks() into synchronous functions, tests don't pass

VladimirZaets commented on src/lib/runChecks.js:
```diff
@@ -0,0 +1,56 @@
+        reporter: 'spec',
+    })
+    const suite = new Mocha.Suite('Validating App Builder templates')
+    suite.timeout(0)
```
Why do we need timeout here? If we have any reason for it, please add the comment to the code

AmyJZhao replied:
This disables the timeout.

VladimirZaets commented on src/lib/runChecks.js:
```diff
@@ -0,0 +1,56 @@
+        )
+    }
+    mocha.suite.addSuite(suite)
+    await new Promise((resolve) => {
```
Can you please explain to me this construction? We add new Promise construction to make synchronous code asynchronous and after that, we await execution to make it back synchronous.

AmyJZhao replied:
that's a really good point. I was intending just the checks to be asynchronous. I removed the Promise construction.

VladimirZaets commented on tests/jest.config.js:
```diff
@@ -0,0 +1,8 @@
+module.exports = {
+    rootDir: '../',
+    testTimeout: 10000, // 10 seconds
```
Is it possible to avoid timeout?

AmyJZhao replied:
I removed the timeout.

AmyJZhao commented:
@VladimirZaets Hi, I submitted a commit that made improvements based on some of your suggestions. I decided to save your other suggestions for a later refactor.
